### DISCUSSION TITLE
Show file data in preview window

### DISCUSSION THREAD

**Discussion by ahrm** [gokcehan/lf#796] [Q&A] [answered]
Hello and thanks for this  great software.

I am using `lf` on windows. Currently when hovering over most non-text files (e.g. images, videos, pdf files, etc.) nothing is shown in the preview window. I am aware that in linux you can use tools such az `useberzug` to get a preview of these files, however, I am not interested in that. All I want is to show some simple meta data about the files in the preview window. For example, file size and dimensions for the image files or the file size and number of pages for the  PDF files.

Is there any way to do this (on windows)?

**Comment by ahrm** [comment, marked as answer]
I wrote a custom previewer to do this (using python). Here is the code in case anybody else wants it:


```
import sys
import mimetypes
import pathlib
import os

def get_mimetype(path):
    return mimetypes.guess_type(path)[0]

def handle_text(path):
    with open(path, 'r') as f:
        print(f.read())

def print_divider():
    print('-' * 80)

def handle_pdf(path):
    try:
        import fitz
        doc = fitz.open(path)
        num_pages = doc.pageCount
        text = clean_text(doc.getPageText(0))
        doc.close()
        print(f'Pages: {num_pages}')
        print(text)
    except ImportError:
        print('Install mupdf for PDF previews')

def clean_text(text):
    # remove all non-ascii characters
    return ''.join(c for c in text if ord(c) < 128)

def handle_image(path):
    try:
        from PIL import Image
        img = Image.open(path)
        width, height = img.size
        print(f'Size: {width}*{height}')
    except ImportError:
        print('Install Pillow for image previews')

def format_time_t(time_t):
    # convert time-t to date
    from datetime import datetime
    return datetime.fromtimestamp(time_t).strftime('%Y-%m-%d %H:%M:%S')

def format_file_size(size_in_bytes):
    if size_in_bytes < 1024:
        return f'{size_in_bytes} B'
    elif size_in_bytes < 1024 * 1024:
        return f'{size_in_bytes / 1024:.2f} KB'
    elif size_in_bytes < 1024 * 1024 * 1024:
        return f'{size_in_bytes / 1024 / 1024:.2f} MB'
    else:
        return f'{size_in_bytes / 1024 / 1024 / 1024:.2f} GB'

if __name__ == '__main__':
    file_path = sys.argv[2]
    mimetype = get_mimetype(file_path)

    if mimetype == None:
        mimetype = 'none'


    try:
        path = pathlib.Path(file_path)
        size = path.stat().st_size
        print("File Size: {}".format(format_file_size(size)))
        print("Modify Time: {}".format(format_time_t(os.path.getmtime(file_path))))


        if mimetype == 'none':
            # if file size is less 100kb
            if size < 1024 * 100:
                print_divider()
                handle_text(file_path)
        if mimetype.startswith('text'):
            print_divider()
            handle_text(file_path)
        if mimetype.startswith('application/pdf'):
            print_divider()
            handle_pdf(file_path)
        if mimetype.startswith('image'):
            print_divider()
            handle_image(file_path)
    except Exception as e:
        print(e)

```

**Reply by gokcehan** [reply]
@ahrm That looks great, thank you. Feel free to add it somewhere in our wiki pages too if you like. Or you might as well consider maintaining it as a separate project. There's [pistol](https://github.com/doronbehar/pistol) that I'm aware of but it may not be working in Windows I think.

By the way, I'm not sure if you are aware but we have an `info` option to be able to display size and time information next to filenames, though displaying it in preview pane also sounds reasonable.

Also, I think this script can help some of the folks in #644 as it seems to filter out non-ascii characters.

**Reply by ahrm** [reply]
I wrote a [blog post](https://medium.com/@a.hr.mostafavi/using-lf-file-manager-on-windows-fc4f1e4e1442) detailing some of the configurations and modifications that I have done. I don't think they are ready for wiki page because they are pretty hacky and contain some hard-coded paths.
You are of course welcome to copy the sections that you deem fit to your wiki (even without attribution).
Cheers.
